File: vcs/benchmarks/baselines/vs_go/allocation.go
// Allocation Baseline - Go Implementation
//
// Measures allocation performance in Go as a baseline for Verum.
//
// Run: go build -o allocation allocation.go && ./allocation

package main

import (
	"fmt"
	"time"
)

const (
	smallIterations  = 1000000
	mediumIterations = 100000
	largeIterations  = 1000
)

type SmallStruct struct {
	A int32
	B int32
	C float64
	D bool
}

//go:noinline
func sink(x interface{}) {
	_ = x
}

func benchmarkSmallAlloc() {
	for i := 0; i < smallIterations; i++ {
		obj := &SmallStruct{
			A: 1,
			B: 2,
			C: 3.14,
			D: true,
		}
		sink(obj)
	}
}

func benchmarkMediumAlloc() {
	for i := 0; i < mediumIterations; i++ {
		data := make([]byte, 1024)
		sink(data)
	}
}

func benchmarkLargeAlloc() {
	for i := 0; i < largeIterations; i++ {
		data := make([]byte, 1024*1024)
		sink(data)
	}
}

func benchmarkSliceGrowth() {
	for round := 0; round < 1000; round++ {
		var s []int32
		for i := 0; i < 10000; i++ {
			s = append(s, int32(i))
		}
		sink(s)
	}
}

func benchmarkSliceWithCapacity() {
	for round := 0; round < 1000; round++ {
		s := make([]int32, 0, 10000)
		for i := 0; i < 10000; i++ {
			s = append(s, int32(i))
		}
		sink(s)
	}
}

func benchmarkMapAlloc() {
	for round := 0; round < 1000; round++ {
		m := make(map[string]int)
		for i := 0; i < 1000; i++ {
			m[fmt.Sprintf("key_%d", i)] = i
		}
		sink(m)
	}
}

func benchmarkChannelAlloc() {
	for i := 0; i < smallIterations/10; i++ {
		ch := make(chan int, 10)
		sink(ch)
	}
}

func main() {
	// Warmup
	benchmarkSmallAlloc()

	// Small allocation
	start := time.Now()
	benchmarkSmallAlloc()
	elapsed := time.Since(start)
	nsPerOp := float64(elapsed.Nanoseconds()) / float64(smallIterations)
	fmt.Printf("Small alloc (struct):      %.2f ns/op\n", nsPerOp)

	// Medium allocation
	start = time.Now()
	benchmarkMediumAlloc()
	elapsed = time.Since(start)
	nsPerOp = float64(elapsed.Nanoseconds()) / float64(mediumIterations)
	fmt.Printf("Medium alloc (1KB slice):  %.2f ns/op\n", nsPerOp)

	// Large allocation
	start = time.Now()
	benchmarkLargeAlloc()
	elapsed = time.Since(start)
	nsPerOp = float64(elapsed.Nanoseconds()) / float64(largeIterations)
	fmt.Printf("Large alloc (1MB slice):   %.2f ns/op\n", nsPerOp)

	// Slice growth
	start = time.Now()
	benchmarkSliceGrowth()
	elapsed = time.Since(start)
	nsPerOp = float64(elapsed.Nanoseconds()) / float64(1000*10000)
	fmt.Printf("Slice growth (append):     %.2f ns/op\n", nsPerOp)

	// Slice with capacity
	start = time.Now()
	benchmarkSliceWithCapacity()
	elapsed = time.Since(start)
	nsPerOp = float64(elapsed.Nanoseconds()) / float64(1000*10000)
	fmt.Printf("Slice with capacity:       %.2f ns/op\n", nsPerOp)

	// Map allocation
	start = time.Now()
	benchmarkMapAlloc()
	elapsed = time.Since(start)
	nsPerOp = float64(elapsed.Nanoseconds()) / float64(1000*1000)
	fmt.Printf("Map insert:                %.2f ns/op\n", nsPerOp)

	// Channel allocation
	start = time.Now()
	benchmarkChannelAlloc()
	elapsed = time.Since(start)
	nsPerOp = float64(elapsed.Nanoseconds()) / float64(smallIterations/10)
	fmt.Printf("Channel alloc:             %.2f ns/op\n", nsPerOp)
}

File: vcs/benchmarks/baselines/vs_rust/allocation.rs
//! Allocation Baseline - Rust Implementation
//!
//! Measures allocation performance in Rust as a baseline for Verum.
//!
//! Compile: rustc -O -o allocation allocation.rs

use std::hint::black_box;
use std::time::Instant;

const SMALL_ITERATIONS: usize = 1_000_000;
const MEDIUM_ITERATIONS: usize = 100_000;
const LARGE_ITERATIONS: usize = 1_000;

#[derive(Clone)]
struct SmallStruct {
    a: i32,
    b: i32,
    c: f64,
    d: bool,
}

fn benchmark_small_alloc() {
    for _ in 0..SMALL_ITERATIONS {
        let obj = Box::new(SmallStruct {
            a: 1,
            b: 2,
            c: 3.14,
            d: true,
        });
        black_box(obj);
    }
}

fn benchmark_medium_alloc() {
    for _ in 0..MEDIUM_ITERATIONS {
        let data: Box<[u8; 1024]> = Box::new([0u8; 1024]);
        black_box(data);
    }
}

fn benchmark_large_alloc() {
    for _ in 0..LARGE_ITERATIONS {
        let data: Vec<u8> = vec![0u8; 1024 * 1024];
        black_box(data);
    }
}

fn benchmark_vec_growth() {
    for _ in 0..1000 {
        let mut v: Vec<i32> = Vec::new();
        for i in 0..10_000 {
            v.push(i);
        }
        black_box(v);
    }
}

fn benchmark_vec_with_capacity() {
    for _ in 0..1000 {
        let mut v: Vec<i32> = Vec::with_capacity(10_000);
        for i in 0..10_000 {
            v.push(i);
        }
        black_box(v);
    }
}

fn benchmark_rc_alloc() {
    use std::rc::Rc;

    for _ in 0..SMALL_ITERATIONS {
        let rc = Rc::new(SmallStruct {
            a: 1,
            b: 2,
            c: 3.14,
            d: true,
        });
        let _clone1 = Rc::clone(&rc);
        let _clone2 = Rc::clone(&rc);
        black_box(rc);
    }
}

fn benchmark_arc_alloc() {
    use std::sync::Arc;

    for _ in 0..SMALL_ITERATIONS {
        let arc = Arc::new(SmallStruct {
            a: 1,
            b: 2,
            c: 3.14,
            d: true,
        });
        let _clone1 = Arc::clone(&arc);
        let _clone2 = Arc::clone(&arc);
        black_box(arc);
    }
}

fn main() {
    // Warmup
    benchmark_small_alloc();

    // Small allocation
    let start = Instant::now();
    benchmark_small_alloc();
    let elapsed = start.elapsed();
    let ns_per_op = elapsed.as_nanos() as f64 / SMALL_ITERATIONS as f64;
    println!("Small alloc (Box):         {:.2} ns/op", ns_per_op);

    // Medium allocation
    let start = Instant::now();
    benchmark_medium_alloc();
    let elapsed = start.elapsed();
    let ns_per_op = elapsed.as_nanos() as f64 / MEDIUM_ITERATIONS as f64;
    println!("Medium alloc (1KB Box):    {:.2} ns/op", ns_per_op);

    // Large allocation
    let start = Instant::now();
    benchmark_large_alloc();
    let elapsed = start.elapsed();
    let ns_per_op = elapsed.as_nanos() as f64 / LARGE_ITERATIONS as f64;
    println!("Large alloc (1MB Vec):     {:.2} ns/op", ns_per_op);

    // Vec growth
    let start = Instant::now();
    benchmark_vec_growth();
    let elapsed = start.elapsed();
    let ns_per_op = elapsed.as_nanos() as f64 / (1000 * 10_000) as f64;
    println!("Vec growth (push):         {:.2} ns/op", ns_per_op);

    // Vec with capacity
    let start = Instant::now();
    benchmark_vec_with_capacity();
    let elapsed = start.elapsed();
    let ns_per_op = elapsed.as_nanos() as f64 / (1000 * 10_000) as f64;
    println!("Vec with_capacity:         {:.2} ns/op", ns_per_op);

    // Rc allocation
    let start = Instant::now();
    benchmark_rc_alloc();
    let elapsed = start.elapsed();
    let ns_per_op = elapsed.as_nanos() as f64 / SMALL_ITERATIONS as f64;
    println!("Rc alloc + 2 clones:       {:.2} ns/op", ns_per_op);

    // Arc allocation
    let start = Instant::now();
    benchmark_arc_alloc();
    let elapsed = start.elapsed();
    let ns_per_op = elapsed.as_nanos() as f64 / SMALL_ITERATIONS as f64;
    println!("Arc alloc + 2 clones:      {:.2} ns/op", ns_per_op);
}
